### PR TITLE
independently trigger responsive mode for responsive components

### PR DESCRIPTION
**Pull request by dms1lva**
Early PR.

Responsive components go in responsive mode only if they themselves need to.
The tab-section moves on a pixel breakpoint

TODO:

- add options to allow customization of responsive triggers

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

---
This change is [Reviewable](https://reviewable.io/reviews/coveo/search-ui/166)

### CODE REVIEW

**Comment by olamothe** [comment]
:lgtm:

I know we have a story for this, but this all seriously needs more UT :) 

---

Reviewed 15 of 15 files at r1.
Review status: 14 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/ui/Facet/FacetSearch.ts, line 58 at r1](https://reviewable.io:443/reviews/coveo/search-ui/166#-KRZBjMHIRpFVKQhYhVf:-KRZBjMHIRpFVKQhYhVg:bzf94so) ([raw file](https://github.com/coveo/search-ui/blob/429ad10b82c91235dc79b6d33c1e087c18e4fe25/src/ui/Facet/FacetSearch.ts#L58)):*
> ```XML
>       // These would point to random instance of a test karma object, and not a real search interface.
>       if (this.facet instanceof Facet && this.facet.searchInterface instanceof SearchInterface) {
>         if (!this.isMobileDevice() && !ResponsiveComponentsUtils.isSmallFacetActivated($$(this.root)) && $$(this.facet.element).hasClass('coveo-facet-searching')) {
> ```

One thing I like to do when there is a lot of conditions like this in a if is extract that into a private method with a relevant name that explains what it's trying to do, with accompanying comment in the method if needed.

For example `if(this.isSmallFacetSearching())` 

---

*[src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts, line 76 at r1](https://reviewable.io:443/reviews/coveo/search-ui/166#-KRZCXI8KmieiLsgGulf:-KRZCXI91eD6stumMWd8:bl8ww27) ([raw file](https://github.com/coveo/search-ui/blob/429ad10b82c91235dc79b6d33c1e087c18e4fe25/src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts#L76)):*
> ```XML
>     this.resizeListener = _.debounce(() => {
>       _.each(this.responsiveComponents, responsiveComponent => {
>         if (this.needTabSection() || this.searchBoxElement && this.tabSection && this.coveoRoot.width() <= ResponsiveComponentsUtils.MEDIUM_MOBILE_WIDTH && this.isFacetActivated) {
> ```

Same comment as before : Try to extract complicated boolean check into a private method with a relevant name.

You can also add some high level comments that explain what adding those class or inserting elements at this place will do in the UI (eg : `Here we're trying to position the tabs under the search box when were in small mode`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/coveo/search-ui/166#-:-KRZD1EoGTXv_Rl-yU9W:b3y57f1)*

**Comment by dms1lva** [comment]
Review status: 8 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/ui/Facet/FacetSearch.ts, line 58 at r1](https://reviewable.io:443/reviews/coveo/search-ui/166#-KRZBjMHIRpFVKQhYhVf:-KRhy76gCMF97j_lEsLP:b-896fix) ([raw file](https://github.com/coveo/search-ui/blob/429ad10b82c91235dc79b6d33c1e087c18e4fe25/src/ui/Facet/FacetSearch.ts#L58)):*
<details><summary><i>Previously, olamothe (Olivier Lamothe) wrote…</i></summary>
> One thing I like to do when there is a lot of conditions like this in a if is extract that into a private method with a relevant name that explains what it's trying to do, with accompanying comment in the method if needed.
> 
> For example `if(this.isSmallFacetSearching())` 
> 
</details>
Done.

---

*[src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts, line 76 at r1](https://reviewable.io:443/reviews/coveo/search-ui/166#-KRZCXI8KmieiLsgGulf:-KRhy8owabagH5q2H5Bf:b-896fix) ([raw file](https://github.com/coveo/search-ui/blob/429ad10b82c91235dc79b6d33c1e087c18e4fe25/src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts#L76)):*
<details><summary><i>Previously, olamothe (Olivier Lamothe) wrote…</i></summary>
> Same comment as before : Try to extract complicated boolean check into a private method with a relevant name.
> 
> You can also add some high level comments that explain what adding those class or inserting elements at this place will do in the UI (eg : `Here we're trying to position the tabs under the search box when were in small mode`)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/coveo/search-ui/166)*

File: sass/_ResponsiveFacets.scss [hunk 1]
```diff
@@ -1,55 +1,46 @@
 @import "Variables.scss";
```
Comment (olamothe) [inline sass/_ResponsiveFacets.scss]: The alignment in this whole file seems bad.

2 spaces everywhere

File: sass/_ResponsiveTabs.scss [hunk 2]
```diff
@@ -1,136 +1,60 @@
- @import "_Variables.scss";
```
Comment (olamothe) [inline sass/_ResponsiveTabs.scss]: Same.

File: src/ui/Facet/Facet.ts [hunk 3]
```diff
@@ -385,7 +387,22 @@ export class Facet extends Component {
      */
     valueCaption: ComponentOptions.buildCustomOption<IStringMap<string>>(() => {
       return null;
-    })
+    }),
+    /**
+     * Specifies if the responsive mode should be enabled on the facets. Responsive mode will make the facet dissapear and instead be
+     * availaible using a dropdown button. Responsive facets are enabled when the width of the element the search interface is bound to
```
Comment (olamothe) [inline src/ui/Facet/Facet.ts]: Remove the "is bound to"

File: src/ui/Facet/Facet.ts [hunk 4]
```diff
@@ -385,7 +387,22 @@ export class Facet extends Component {
...
+     * reaches 800 pixels. This value can be modified using {@link Facet..optionsresponsiveBreakpoint}. Disabling reponsive mode for one tab will disable it for all tabs.
```
Comment (olamothe) [inline src/ui/Facet/Facet.ts]: this link will be broken. It needs to be Facet.options.responsiveBreakpoint

Comment (olamothe) [inline src/ui/Facet/Facet.ts]: Replace "tab" with "facet". A bit too good with the copy paste :)

File: src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts [hunk 5]
```diff
@@ -32,22 +33,24 @@ export class ResponsiveComponentsManager {
...
   private responsiveFacets: ResponsiveFacets;
+  private tabSectionPreviousSibling: Dom;
+  private tabSectionParent: Dom;
+  private disabledComponents: string[] = [];
 
   // Register takes a class and will instantiate it after framework initialization has completed.
-  public static register(responsiveComponentConstructor: IResponsiveComponentConstructor, root: Dom, ID: string, component: IComponentDefinition) {
-    let searchInterface = <SearchInterface>Component.get(root.el, SearchInterface, true);
+  public static register(responsiveComponentConstructor: IResponsiveComponentConstructor, root: Dom, ID: string,
+    component: Component, options: IResponsiveComponentOptions): void {
```
Comment (olamothe) [inline src/ui/ResponsiveComponents/ResponsiveComponentsManager.ts]: Put the constructor signature all on the same line.

File: src/ui/Tab/Tab.ts [hunk 6]
```diff
@@ -119,7 +120,16 @@ export class Tab extends Component {
      * If cached results are available but are older than the specified age, a new query will be performed on the index.<br/>
      * By default, this is left undefined and the Coveo Search API will decide the cache duration.
      */
-    maximumAge: ComponentOptions.buildNumberOption()
+    maximumAge: ComponentOptions.buildNumberOption(),
+    /**
+     * Specifies if the responsive mode should be enabled for the tabs. Responsive mode will make the overflowing tabs dissapear and instead
+     * be availaible using a dropdown button. Responsive tabs are enabled when tabs overflow or when the width of the search interface
+     * becomes too small. Disabling reponsive mode for one tab will disable it for all tabs.
+     * Therefore, this options only needs to be set on one tab to be effective.
```
Comment (olamothe) [inline src/ui/Tab/Tab.ts]: I would add 2 new line between paragraph to make it easier to read in the generated doc.

File: sass/_ResponsiveFacets.scss (hunk 1 above)
Comment (dms1lva) [inline sass/_ResponsiveFacets.scss]: Done

File: sass/_ResponsiveTabs.scss (hunk 2 above)
Comment (dms1lva) [inline sass/_ResponsiveTabs.scss]: Done

**Pull request merged**